packages/opentelemetry-plugin-express/src/express.ts

```diff
@@ -0,0 +1,220 @@
+/*!
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { BasePlugin } from '@opentelemetry/core';
+import { Attributes } from '@opentelemetry/types';
+import * as express from 'express';
+import * as core from 'express-serve-static-core';
+import * as shimmer from 'shimmer';
+import {
+  ExpressLayer,
+  ExpressRouter,
+  AttributeNames,
+  PatchedRequest,
+  Parameters,
+  PathParams,
+  _LAYERS_STORE_PROPERTY,
+  ExpressPluginConfig,
+  ExpressLayerType,
+} from './types';
+import {
+  getLayerMetadata,
+  storeLayerPath,
+  patchEnd,
+  isLayerIgnored,
+} from './utils';
+import { VERSION } from './version';
+
+/**
+ * This symbol is used to mark express layer as being already instrumented
+ * since its possible to use a given layer multiple times (ex: middlewares)
+ */
+export const kLayerPatched: unique symbol = Symbol('express-layer-patched');
+
+/** Express instrumentation plugin for OpenTelemetry */
+export class ExpressPlugin extends BasePlugin<typeof express> {
+  readonly _COMPONENT = 'express';
+  readonly supportedVersions = ['^4.0.0'];
+  protected _config!: ExpressPluginConfig;
+
+  constructor(readonly moduleName: string) {
+    super('@opentelemetry/plugin-express', VERSION);
+  }
+
+  /**
+   * Patches Express operations.
+   */
+  protected patch() {
+    this._logger.debug('Patching Express');
+
+    if (this._moduleExports === undefined || this._moduleExports === null) {
+      return this._moduleExports;
+    }
+    const routerProto = (this._moduleExports
+      .Router as unknown) as express.Router;
```

Comment:
Is `typeof this._moduleExports` not automatically `express`?

Reply:
It is, but `this._moduleExports.Router` return a constructor which for a mysterious reason `wrap` returns a type error because it cant find methods on it. I could use `this._moduleExports.Router.prototype` but it doesnt work at runtime :/